omp/solver/idr_kernels.cpp

```diff
@@ -0,0 +1,348 @@
+            norm += squared_norm(subspace_vectors->at(row, j));
+        }
+
+        norm = sqrt(norm);
+
+#pragma omp parallel for
+        for (size_type j = 0; j < num_cols; j++) {
+            subspace_vectors->at(row, j) /= norm;
+        }
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(GKO_DECLARE_IDR_INITIALIZE_KERNEL);
+
+
+template <typename ValueType>
+void step_1(std::shared_ptr<const OmpExecutor> exec, const size_type k,
+            const matrix::Dense<ValueType> *m,
+            const matrix::Dense<ValueType> *f,
+            const matrix::Dense<ValueType> *residual,
+            const matrix::Dense<ValueType> *g, matrix::Dense<ValueType> *c,
+            matrix::Dense<ValueType> *v,
+            const Array<stopping_status> *stop_status)
+{
+    const auto m_size = m->get_size();
+    const auto nrhs = f->get_size()[1];
+
+    // Compute c = M \ f
+    solve_lower_triangular(m, f, c, stop_status);
+
+    for (size_type i = 0; i < nrhs; i++) {
+        if (stop_status->get_const_data()[i].has_stopped()) {
+            continue;
+        }
+
+        // v = residual - c_k * g_k - ... - c_s * g_s
+#pragma omp parallel for
+        for (size_type row = 0; row < v->get_size()[0]; row++) {
+            auto temp = residual->at(row, i);
+            for (size_type j = k; j < m->get_size()[0]; j++) {
+                temp -= c->at(j, i) * g->at(row, j * nrhs + i);
+            }
+            v->at(row, i) = temp;
+        }
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(GKO_DECLARE_IDR_STEP_1_KERNEL);
+
+
+template <typename ValueType>
+void step_2(std::shared_ptr<const OmpExecutor> exec, const size_type k,
+            const matrix::Dense<ValueType> *omega,
+            const matrix::Dense<ValueType> *preconditioned_vector,
+            const matrix::Dense<ValueType> *c, matrix::Dense<ValueType> *u,
+            const Array<stopping_status> *stop_status)
+{
+    const auto nrhs = omega->get_size()[1];
+    for (size_type i = 0; i < nrhs; i++) {
+        if (stop_status->get_const_data()[i].has_stopped()) {
+            continue;
+        }
+
+#pragma omp parallel for
+        for (size_type row = 0; row < u->get_size()[0]; row++) {
+            auto temp = omega->at(0, i) * preconditioned_vector->at(row, i);
+            for (size_type j = k; j < c->get_size()[0]; j++) {
+                temp += c->at(j, i) * u->at(row, j * nrhs + i);
+            }
+            u->at(row, k * nrhs + i) = temp;
+        }
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(GKO_DECLARE_IDR_STEP_2_KERNEL);
+
+
+template <typename ValueType>
+void step_3(std::shared_ptr<const OmpExecutor> exec, const size_type k,
+            const matrix::Dense<ValueType> *p, matrix::Dense<ValueType> *g,
+            matrix::Dense<ValueType> *g_k, matrix::Dense<ValueType> *u,
+            matrix::Dense<ValueType> *m, matrix::Dense<ValueType> *f,
+            matrix::Dense<ValueType> *alpha, matrix::Dense<ValueType> *residual,
```

Comment:
Since you don't use alpha here, you can remove its name for clarity. Same for reference.